mbuild/utils/specific_ff_to_residue.py

```diff
@@ -0,0 +1,385 @@
+        structure's order.  If True, the residues will be reordered as they appear in the residues
+        variable.  If False, the order will be the same as entered in the original structure.
+    box: list, default=None
+        list of 3 positive float or integer values or the dimensions [x, y ,z]
+        for structure in nanometers (nm). This is to add/override or change the structures dimensions.
+        Ex: [1,2,3]
+    boxes_for_simulation: int [1, 2], default = 1
+        Gibbs (GEMC) or grand canonical (GCMC) ensembles are examples of where the boxes_for_simulation would be 2.
+        Canonical (NVT) or isothermal–isobaric (NPT) ensembles are example with the boxes_for_simulation equal to 1.
+        Note: the only valid options are 1 or 2.
+
+    Returns
+    -------
+    list, [structure, coulomb14scalar_dict, lj14_scalar_dict, residues_applied_list]
+        structure: parmed.Structure
+            parmed structure with applied force field
+        coulomb14scalar_dict: dict
+            a dictionary with the 1,4-colombic scalars for each residue
+                (i.e., a different force field could on each residue)
+        lj14_scalar_dict: dict
+            a dictionary with the 1,4-LJ scalars for each residue
+            (i.e., a different force field could on each residue)
+        residues_applied_list: list
+            list of residues (i.e., list of stings).
+            These are all the residues in which the force field actually applied
+
+    Notes
+    -----
+    To write the NAMD/GOMC force field, pdb, psf, and force field
+    (.inp) files, the residues and forcefields must be provided in
+    a str or dictionary. If a dictionary is provided all residues must
+    be specified to a force field if the boxes_for_simulation is equal to 1.
+
+    Generating an empty box (i.e., pdb and psf files):
+    Enter residues = [], but the accompanying structure must be an empty mb.Box.
+    However, when doing this, the forcefield_selection must be supplied,
+    or it will provide an error (i.e., forcefield_selection can not be equal to None).
+
+    In this current FF/psf/pdb writer, a residue type is essentially a molecule type.
+    Therefore, it can only correctly write systems where every bead/atom in the molecule
+    has the same residue name, and the residue name is specific to that molecule type.
+    For example: a protein molecule with many residue names is not currently supported,
+    but is planned to be supported in the future.
+    """
+
+    if has_foyer:
+        from foyer import Forcefield
+        from foyer.forcefields import forcefields
+    else:
+        print_error_message = 'Package foyer is not installed. '\
+                              'Please install it using conda install -c conda-forge foyer'
+        raise ImportError(print_error_message)
+
+    if not isinstance(structure, Compound) and not isinstance(structure, mb.Box):
```

Comment:
I think this should be an `or` clause, not `and`. Python's [`and` and `or` operators are short-circuit-ing ones](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not), so if the object is _not_ a `Compound`, it will not evaluate the second part of this condition, since the `and` operator ensures it will always be false.

Reply:
I believe "and" is the proper statement.  If you put "or" in place of "and", you end up with it failing for when it is a Compound (see the below error).

TypeError: ERROR: The structure expected to be of type: <class 'mbuild.compound.Compound'> or <class 'mbuild.box.Box'>, received: <class 'mbuild.compound.Compound'>

Reply:
Actually, there is an even better way presented to me by @umesh-timalsina and @daico007 . [`isinstance` can take a tuple of types to check against](https://docs.python.org/3/library/functions.html#isinstance)

```python3
if not isinstance(structure, (Compound, mb.Box)):
....

Reply:
As requested I change this to if "not isinstance(structure, (Compound, mb.Box)):" and similar code in the charmm_writer to the same format.